docs/proposals/20210335-load-balancer-provider.md

```diff
@@ -0,0 +1,999 @@
+
+### Requirements
+
+#### Functional Requirements
+
+##### FR1
+
+The Load Balancer provider WILL NOT automate the creation of DNS records, as this is deferred to an external provider. 
+
+##### FR2
+
+The Load Balancer provider MUST not exclude the use of external controllers to manage DNS for load balancers.
+
+#### Non-Functional Requirements
+
+##### NFR1
+
+The Load Balancer MUST allow alternate endpoints, such as DNS aliases to be used as the control plane endpoint in addition to the load balancer.
+
+##### NFR2
+
+This proposal MUST be compatible with the [Cluster API Provider AWS v1alpha4 API](https://docs.google.com/document/d/1EeD2KjFfeKoD-o0LJTMMWh9GUy-15yG80S5Uoj7oKcA/edit)
+
+### Implementation Details/Notes/Constraints
+
+#### High level Sequence Diagrams
+
+![With Load Balancer Provider](images/load-balancer/cluster-creation-w-loadbalancer-sequence.png)
+
+![Without Load Balancer Provider](images/load-balancer/cluster-creation-wo-loadbalancer-sequence.png)]
+
+### Changes to types
+
+#### Changes to Cluster type:
+
+Consolidate Control Plane related configuration under `Spec.ControlPlane`.
+
+- Create a new field Spec.ControlPlane
+    - Move Spec.ControlPlaneEndpoint to Spec.ControlPlane.Endpoint
+    - Move Spec.ControlPlaneRef to Spec.ControlPlane.ProviderRef
+
+```yaml
+apiVersion: cluster.x-k8s.io/v1alpha4
+kind: Cluster
+metadata:
+  name: my-cluster
+spec:
+  controlPlane:
+    endpoint:
+      host: test.example.com
+      port: 443
+    providerRef:
+      apiGroup: cluster.x-k8s.io
+      kind: KubeadmControlPlane
+      name: my-cluster-control-plane
+```
+
+#### Changes to Cluster reconciliation:
+
+![Cluster Reconciliation Flow](images/load-balancer/cluster-reconciliation.png)
+
+##### Scenario: externally managed endpoint and control plane
+
+User wants to provide an externally managed endpoint for use when creating a cluster to a pre-existing control plane that is not managed by Cluster API.
+
+Preconditions:
+- `Spec.ControlPlane.Endpoint` is non-empty
+- `Spec.ControlPlane.ProviderRef` is nil
+- A pre-configured kubeconfig secret that can access the pre-existing control plane exists with the expected name.
+
+Reconciliation:
+- Set `ControlPlaneInitializedCondition` to True
+- Test availability of the controlplane using the configured endpoint and provided kubeconfig:
+  - If available: Set `Status.ControlPlaneReady` and the `ControlPlaneReadyCondition` to True
+  - Otherwise: Set `Status.ControlPlaneReady` and the `ControlPlaneReadyCondition` to False
+
+##### Scenario: externally managed endpoint
+
+User wants to provide an externally managed endpoint for use when creating a cluster, but wants Cluster API to manage the control plane for the cluster.
+
+Preconditions:
+- `Spec.ControlPlane.Endpoint` is non-empty
+- `Spec.ControlPlane.ProviderRef` is defined
+
+Reconciliation:
+- Fetch the referenced provider CR as `cp`
+  - If `cp.Spec.Endpoint` is empty, set `cp.Spec.Endpoint` to `Spec.ControlPlane.Endpoint`
```

Comment:
Might be we should point out that `cp.Spec.Endpoint` is a new requirement before introducing it here without further context